rust/datafusion/src/datasource/datasource.rs

```diff
@@ -81,6 +81,7 @@ pub trait TableProvider {
         projection: &Option<Vec<usize>>,
         batch_size: usize,
         filters: &[Expr],
+        limit: Option<usize>,
```

Comment:
I recommend adding a comment somewhere here saying that the `limit` is advisory or something -- namely datafusion does not rely on it for correctness,  rather that datafusion will only use that many rows.